<a href="https://colab.research.google.com/github/NataliaGon/kpi/blob/AI-cybersecurity/Lab9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import numpy as np
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
from sklearn.metrics import classification_report


(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Нормалізація пікселів до [0,1]
x_train = x_train.astype("float32") / 255.0
x_test = x_test.astype("float32") / 255.0

# Перетворюємо мітки у one-hot формат (10 класів)
y_train_cat = to_categorical(y_train, num_classes=10)
y_test_cat = to_categorical(y_test, num_classes=10)


0.0


# Побудова нейронної мережі (MLP)

In [3]:
model = Sequential([
    Flatten(input_shape=(28, 28)),     # 28x28 -> 784
    Dense(128, activation="relu"),     # прихований шар
    Dense(64, activation="relu"),      # ще один прихований шар (можна забрати)
    Dense(10, activation="softmax")    # вихідний шар, 10 класів
])



/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


# Компіляція моделі, функція втрат + оптимізатор + метрики

In [16]:
model.compile(
    loss="categorical_crossentropy",
    optimizer="adam",
    metrics=["accuracy"]
)

# Навчання моделі

In [17]:
history = model.fit(
    x_train, y_train_cat,
    batch_size=128,
    epochs=10,
    validation_split=0.1,
    verbose=1
)



Epoch 1/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.9937 - loss: 0.0204 - val_accuracy: 0.9777 - val_loss: 0.0943
Epoch 2/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9948 - loss: 0.0177 - val_accuracy: 0.9770 - val_loss: 0.0963
Epoch 3/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9947 - loss: 0.0179 - val_accuracy: 0.9762 - val_loss: 0.0956
Epoch 4/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9963 - loss: 0.0134 - val_accuracy: 0.9778 - val_loss: 0.0912
Epoch 5/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9969 - loss: 0.0110 - val_accuracy: 0.9785 - val_loss: 0.0978
Epoch 6/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9970 - loss: 0.0095 - val_accuracy: 0.9770 - val_loss: 0.1185
Epoch 7/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9965 - loss: 0.0110 - val_accuracy: 0.9803 - val_loss: 0.0940
Epoch 8/10
422/422 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9983 - loss: 0.0061 - val_accuracy: 0.

# Оцінка моделі на тестових даних

In [18]:
test_loss, test_acc = model.evaluate(x_test, y_test_cat, verbose=0)
print(f"Test loss: {test_loss:.4f}")
print(f"Test accuracy: {test_acc:.4f}")

y_pred_probs = model.predict(x_test)
y_pred = np.argmax(y_pred_probs, axis=1)

print(classification_report(y_test, y_pred, digits=4))



Test loss: 0.0916
Test accuracy: 0.9803
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step
              precision    recall  f1-score   support

           0     0.9878    0.9878    0.9878       980
           1     0.9921    0.9903    0.9912      1135
           2     0.9694    0.9835    0.9764      1032
           3     0.9880    0.9782    0.9831      1010
           4     0.9825    0.9745    0.9785       982
           5     0.9832    0.9843    0.9838       892
           6     0.9751    0.9823    0.9787       958
           7     0.9843    0.9737    0.9790      1028
           8     0.9666    0.9805    0.9735       974
           9     0.9731    0.9673    0.9702      1009

    accuracy                         0.9803     10000
   macro avg     0.9802    0.9802    0.9802     10000
weighted avg     0.9804    0.9803    0.9803     10000



# Висновки

Отримані результати свідчать про високу якість моделі: точність на тестовому наборі перевищує 97 %, а значення precision, recall та F1-показника залишаються стабільно високими для всіх класів. Мережа продемонструвала здатність коректно узагальнювати інформацію та ефективно розпізнавати нові зображення, що підтверджує адекватність обраної архітектури, правильність підготовки даних та ефективність навчання. Отже, створена модель успішно виконує поставлене завдання та може бути використана як базове рішення для задач класифікації зображень рукописних цифр.